pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go

```diff
@@ -913,30 +936,79 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
 	}
 
 	if lbaas.opts.ManageSecurityGroups {
+		err := lbaas.ensureSecurityGroup(clusterName, apiService, nodes, loadbalancer)
+		if err != nil {
+			// cleanup what was created so far
+			_ = lbaas.EnsureLoadBalancerDeleted(clusterName, apiService)
+			return status, err
+		}
+	}
+
+	return status, nil
+}
+
+// ensureSecurityGroup ensures security group exist for specific loadbalancer service.
+// Creating security group for specific loadbalancer service when it does not exist.
+func (lbaas *LbaasV2) ensureSecurityGroup(clusterName string, apiService *v1.Service, nodes []*v1.Node, loadbalancer *loadbalancers.LoadBalancer) error {
+	// find node-security-group for service
+	var err error
+	if len(lbaas.opts.NodeSecurityGroupIDs) == 0 {
+		lbaas.opts.NodeSecurityGroupIDs, err = getNodeSecurityGroupIDForLB(lbaas.compute, nodes)
+		if err != nil {
+			glog.Errorf("Failed to find node-security-group for loadbalancer service %s/%s: %v", apiService.Namespace, apiService.Name, err)
+			return fmt.Errorf("Failed to find node-security-group for loadbalancer service %s/%s: %v", apiService.Namespace, apiService.Name, err)
+		}
+	}
+	glog.V(4).Infof("find node-security-group %v for loadbalancer service %s/%s", lbaas.opts.NodeSecurityGroupIDs, apiService.Namespace, apiService.Name)
+
+	// get service ports
+	ports := apiService.Spec.Ports
+	if len(ports) == 0 {
+		return fmt.Errorf("no ports provided to openstack load balancer")
+	}
+
+	// get service source ranges
+	sourceRanges, err := service.GetLoadBalancerSourceRanges(apiService)
+	if err != nil {
+		return fmt.Errorf("Failed to get source ranges for loadbalancer service %s/%s: %v", apiService.Namespace, apiService.Name, err)
+	}
+
+	// ensure security group for LB
+	lbSecGroupName := getSecurityGroupName(clusterName, apiService)
+	lbSecGroupID, err := groups.IDFromName(lbaas.network, lbSecGroupName)
+	if err != nil {
+		// check whether security group does not exist
+		_, ok := err.(*gophercloud.ErrResourceNotFound)
+		if ok {
+			// create it later
+			lbSecGroupID = ""
+		} else {
+			glog.Errorf("Error occurred finding security group: %s: %v", lbSecGroupName, err)
+			return fmt.Errorf("Error occurred finding security group: %s: %v", lbSecGroupName, err)
+		}
+	}
+	if len(lbSecGroupID) == 0 {
+		// create security group
 		lbSecGroupCreateOpts := groups.CreateOpts{
 			Name:        getSecurityGroupName(clusterName, apiService),
-			Description: fmt.Sprintf("Securty Group for %v Service LoadBalancer", apiService.Name),
+			Description: fmt.Sprintf("Securty Group for loadbalancer service %s/%s", apiService.Namespace, apiService.Name),
 		}
 
 		lbSecGroup, err := groups.Create(lbaas.network, lbSecGroupCreateOpts).Extract()
-
 		if err != nil {
-			// cleanup what was created so far
-			_ = lbaas.EnsureLoadBalancerDeleted(clusterName, apiService)
```

Comment:
all the way through this function, you've removed this call to cleanup during error unwind.  That's fine, but that makes it likely that this function will be called multiple times with some first portion of the resources already existing from a previous failed attempt.  With this new approach, we need to make sure that _every_ call to create a resource will correctly/safely ignore the case where the resource already exists.

See `openstack_routes.go` for an alternative pattern.  The routes code accumulates "undo" callbacks, and calls them in reverse order if one of the steps encounters an error.  (see references to `onFailure`)